opentelemetry-sdk/src/logs/record.rs

```diff
@@ -97,6 +97,28 @@ impl opentelemetry::logs::LogRecord for LogRecord {
     }
 }
 
+impl LogRecord {
+    /// Provides an iterator over the attributes in the `LogRecord`.
+    pub fn attributes_iter(&self) -> impl Iterator<Item = &(Key, AnyValue)> {
+        self.attributes
+            .as_ref()
+            .map_or_else(|| [].iter(), |attrs| attrs.iter())
+    }
+
+    /// Returns the number of attributes in the `LogRecord`.
+    pub fn attributes_len(&self) -> usize {
```

Comment:
This method also need not be made public?

Reply:
Without this, the only way to get the length would be to iterate over all the entries. I can keep it private if we are sure this won't be used by exporters/processors.

Reply:
I would be in favor of keeping the public API surface smaller unless we already have users asking for this feature. This could always be added later if there's a significant demand for it.

Reply:
Also, if we do decide to give users the information about length of the iterator, we should probably do it by overriding the `Iterator` traits methods instead of coming up with our own. That's probably the idiomatic way of doing it.

- [Iterator::size_hint](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.size_hint)
- [Iterator::count](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.count)

Reply:
[Iterator::count](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.count) will iterate over the entries to find the size - not sure if we really want this, as that would have o(n) complexity.

Reply:
Given there is no other performant way to obtain length, I think we'll need to expose attributes length in some way, but lets start without it and add later.

Reply:
I can remove it if we all agree. As of now, it's used in otel-appender-tracing/logs unit-test to validate the attributes, will create the wrapper over the attributes_iter() there.

Reply:
if the only use is in test, lets remove it actually and make a helper method in tests.

Reply:
> [Iterator::count](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.count) will iterate over the entries to find the size - not sure if we really want this, as that would have o(n) complexity.

Yes, we don't want that. That's the default behavior of the count method. My suggestion was to override these methods and not use them as-is. You should be able to override them to have a more performant implementation.

Reply:
> > [Iterator::count](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.count) will iterate over the entries to find the size - not sure if we really want this, as that would have o(n) complexity.
> 
> Yes, we don't want that. That's the default behavior of the count method. My suggestion was to override these methods and not use them as-is. You should be able to override them to have a more performant implementation.

That is good! I didn't realize we could provide own implementation to return the known size without iterating.

Reply:
> That is good! I didn't realize we could provide own implementation to return the known size without iterating.

We shouldn't be doing that. Any method over Iterator trait should be consuming the iterator, iterating over it and giving the required result.  If we need to provide length, we should be providing a separate method, just like provided by other containers.

Reply:
> Any method over Iterator trait should be consuming the iterator, iterating over it and giving the required result.

Is this a requirement? The docs for [implementing iterator](https://doc.rust-lang.org/std/iter/index.html#implementing-iterator) says that we could write a custom implementation for the iterator methods (without repeatedly calling `next()`):

_"Also note that Iterator provides a default implementation of methods such as nth and fold which call next internally. **However, it is also possible to write a custom implementation of methods like nth and fold if an iterator can compute them more efficiently without calling next.**"_

Reply:
The issue is, any method on the iterator should start processing from the current position of the iterator. So Iter::count() shouldn't be giving the actual size of underlying storage, but the length from the current position.
```rust
fn main() {
    let vec = vec![1, 2, 3, 4, 5];
    let mut iter = vec.iter();
    let mut count = iter.clone().count();
    println!("count: {}", count); // 5

    iter.next();
    count = iter.count();

    println!("new count: {}", count); // 4
}
```